### ISSUE TITLE
UI types should allow to express stacked widgets

### ISSUE THREAD

**Issue by romanofski** [purebred-mua/purebred#297]
**Is your feature request related to a problem? Please describe.**
Brick allows rendering widgets on top of other widgets. That is achieved by rendering all widgets in the background first, and then render the foreground widget on top. For example:

```Haskell
drawUI :: AppState -> [Widget Name]
drawUI s = let ui = vBox (renderWidget s (focusedViewName s) <$> focusedViewWidgets s)
           in case focusedViewWidget s of
                ConfirmDialog -> [renderConfirm s, ui]
                _ -> [ui]
```

As you can see, this "hack" above is used (from #119), because the UI has no way of expressing, that widgets need to be stacked in order to render them on top of each other.

Currently we are rendering a flat list of widgets defined in `UI.Views` and `App.hs`. Each View defines a list of widgets which are `Visible` to be rendered. The flat list allows us to conveniently flip their viewing state in `Actions.hs` whenever we want to hide or display additional widgets.

**Describe the solution you'd like**
The solution needs to use a more expressive data type which allows:
* to allow conveniently manipulate the viewing state, while
* express the stacking order of the widgets

**Describe alternatives you've considered**
Don't know how exactly would work and which data types would be suitable.

**Additional context**
None

**Comment by romanofski** [comment]
I thought about this a bit. Currently we have:
```Haskell
data View = View
    { _vFocus :: Name
    , _vWidgets :: Tiles
    }

-- A tile is a view element with a widget and it's view state
newtype Tiles =
  Tiles (V.Vector Tile)
  deriving (Eq, Show)

type instance Index Tiles = Name
type instance IxValue Tiles = Tile

instance Ixed Tiles where
  ix = tile
```
and I thought, well we could express the stacked nature by making the Tiles a list of lists like:

```Haskell
newtype Tiles =
  Tiles (V.Vector (V.Vector Tile))
  deriving (Eq, Show)
```

(perhaps with a better data type). The view definition would then become from:

```Haskell
composeView :: View
composeView =
  View
    { _vWidgets =
        Tiles $ fromList
          [ Tile Visible ComposeHeaders
          , Tile Visible ComposeListOfAttachments
          , Tile Visible StatusBar
          , Tile Hidden ComposeFrom
          , Tile Hidden ComposeTo
          , Tile Hidden ComposeSubject
          , Tile Hidden ConfirmDialog
          ]
    , _vFocus = ComposeFrom
    }
```
to
```Haskell
composeView :: View
composeView =
  View
    { _vWidgets =
        fromList [
         Tiles $ fromList
          [ Tile Visible ComposeHeaders
          , Tile Visible ComposeListOfAttachments
          , Tile Visible StatusBar
          , Tile Hidden ComposeFrom
          , Tile Hidden ComposeTo
          , Tile Hidden ComposeSubject
         ],
         , Tiles $ fromList [
          Tile Hidden ConfirmDialog
          ]
    , _vFocus = ComposeFrom
    }
```
 Then render "simply" foldMap from left to right, or right to left, whatever helps.

The only tricky thing I've encountered with this approach is that indexing by `Name` will not work anymore, since I have now a two dimensional matrix. Perhaps indexing by `(Int, Name)` might work, but it's cumbersome.

Perhaps the list of lists approach is afterall not as ideal as I initially thought.

**Comment by romanofski** [comment]
I went down this rabbit hole while we're waiting for another purebred-email release and the patches are parked. I called the additional indirection `Layers`
```Haskell
newtype ViewLayers = ViewLayers (V.Vector Tiles)

data View = View
    { _vFocus :: Name
    , _vWidgets :: ViewLayers  -- proof of concept
    }
```
since I thought of it as a stack of layers. I'm not sure if `Tile` is the right name and it should perhaps change.

I got it working today and there is overall not much which needs to change. I also used a `type instance` for `IxValue` to basically index by widget name. Example:
```Haskell
instance Completable 'ComposeTo where
  complete _ = pure . set (asViews . vsViews . at ComposeView . _Just . vWidgets . ix ComposeTo . ix ComposeTo . veState) Hidden
```
The first `ix ComposeTo` picks the right layer and the second `ix ComposeTo` the right `Tile`. I'm now almost certain that this is unnecessary complexity and a normal index by number should be fine. I'm not planning to move widgets between layers, so there shouldn't be much to worry about. Furthermore, picking the right layer by included widget name needs to search through the entire `Vector` which isn't very performant too.
